### PR TITLE
Gerald/local istorage

### PR DESCRIPTION
**Pull request by geralddejong**

### CODE REVIEW

File: client/src/app.tsx [hunk 1]
```diff
@@ -130,18 +133,36 @@ class App extends React.Component<IAppProps, IAppState> {
...
-                </div>
+                ) : (
+                    <h1>No island!</h1>
+                )}
             </div>
         )
     }
+
+    private fetchIsland(islandName: string): void {
+        this.props.storage.getIslandData(islandName).then(islandData => {
```
Comment (mitchelljustin) [inline client/src/app.tsx:154]: This would read nicer if you used async/await instead https://javascript.info/async-await

Comment (geralddejong) [reply client/src/app.tsx:154]: ooh cool yeah that's the thing i wanted to learn

File: client/src/app.tsx [hunk 2]
```diff
@@ -130,18 +133,36 @@ class App extends React.Component<IAppProps, IAppState> {
...
+            if (islandData) {
+                const island = new Island(
+                    this.islandSubject,
+                    islandData,
+                    this.fabricKernel,
+                    this.props.storage,
+                )
+                island.state.dispatch()
```
Comment (mitchelljustin) [inline client/src/app.tsx:162]: What does `.dispatch()` do? Is it a RXjs thing?

File: client/src/body/physics.ts [hunk 3]
```diff
@@ -20,11 +18,20 @@ export interface IPhysicsFeature {
...
 }
 
+function getPhysicsFeature(feature: PhysicsFeature): number {
+    const value = localStorage.getItem(feature)
+    return value ? parseFloat(value) : 1.0
+}
+
+function setPhysicsFeature(feature: PhysicsFeature, factor: number): void {
+    localStorage.setItem(feature, factor.toFixed(3))
+}
```
Comment (mitchelljustin) [inline client/src/body/physics.ts:28]: This is still just for testing purposes right?

File: client/src/index.tsx [hunk 4]
```diff
@@ -6,11 +6,15 @@ import App from "./app"
 import {IFabricExports} from "./body/fabric-exports"
 import "./index.css"
 import registerServiceWorker from "./service-worker"
-import {LocalStorage} from "./storage/local-storage"
+import {BrowserStorage} from "./storage/browser-storage"
+import {RemoteStorage} from "./storage/remote-storage"
+import {IStorage} from "./storage/storage"
 
+const REMOTE_STORAGE = false
```
Comment (mitchelljustin) [inline client/src/index.tsx:13]: There's a way to read this from Node environment variables using Webpack. https://stackoverflow.com/questions/49616291/process-env-on-client-side-reactjs

File: client/src/app.tsx (hunk 2 above)
Comment (geralddejong) [reply client/src/app.tsx:162]: No, it calls next() on a Subject<IslandState>

File: client/src/island/hexalot.ts [hunk 5]
```diff
@@ -69,34 +77,60 @@ export class Hexalot {
...
-        return !!this.genome
-    }
-
     public get isLegal(): boolean {
         return this.spots.every(spot => spot.isLegal)
     }
 
-    public refreshFingerprint(): void {
+    public refreshId(): void {
         this.identifier = spotsToHexFingerprint(this.spots)
```
Comment (mitchelljustin) [inline client/src/island/hexalot.ts]: Change the name of `spotsToHexFingerprint` to ID as well

File: client/src/island/hexalot.ts [hunk 6]
```diff
@@ -69,34 +77,60 @@ export class Hexalot {
...
 
+    public get vacant(): boolean {
+        return this.genomeStatus === LoadStatus.Loaded && !this.genome
+    }
+
+    public fetchGenome(storage: IStorage): boolean {
+        switch (this.genomeStatus) {
+            case LoadStatus.Pending:
+                this.genomeStatus = LoadStatus.Busy
+                storage.getGenomeData(this).then(genomeData => {
```
Comment (mitchelljustin) [inline client/src/island/hexalot.ts:99]: async/await

File: client/src/island/hexalot.ts [hunk 7]
```diff
@@ -69,34 +77,60 @@ export class Hexalot {
...
+            case LoadStatus.Loaded:
+                return true
+        }
+    }
+
+    public fetchJourney(storage: IStorage, island: Island, loaded: () => void): boolean {
+        switch (this.journeyStatus) {
+            case LoadStatus.Pending:
+                this.journeyStatus = LoadStatus.Busy
+                storage.getJourneyData(this).then(journeyData => {
```
Comment (mitchelljustin) [inline client/src/island/hexalot.ts:117]: async/await

File: client/src/island/hexalot.ts [hunk 8]
```diff
@@ -167,22 +200,3 @@ export class Hexalot {
...
-//         return [b0, b1, b2, b3];
-//     });
-//     const landStack = [].concat.apply([], booleanArrays).reverse();
-//     spots.forEach(spot => spot.land = landStack.pop());
-// };
-//
-// export const hexalotFromFingerprint = (fingerprint: string, index: number, hexalotMasterLookup: (fingerprint: string) => string): Hexalot => {
-//     const hexalot = new Hexalot(undefined, {x: 0, y: 0}, GOTCH_SHAPE.map(c => new Spot(c)));
-//     fingerprintToSpots(fingerprint, hexalot.spots);
-//     return hexalot;
```
Comment (mitchelljustin) [inline client/src/island/hexalot.ts:187]: Delete this?

File: client/src/body/physics.ts (hunk 3 above)
Comment (geralddejong) [reply client/src/body/physics.ts:28]: yes, it's not even operative at the moment.  but when we want to have physics constants different for different islands we will have to design a few workable universes with this.

File: client/src/island/island-state-command.ts [hunk 9]
```diff
@@ -190,5 +197,27 @@ export class IslandStateCommand {
 
         }
     }
+
+    private claimHexalot(hexalot: Hexalot, state: IslandState): void {
+        this.state.storage.claimHexalot(this.state.island, hexalot, freshGenome().genomeData)
+            .then(islandPattern => {
+                if (islandPattern) {
```
Comment (mitchelljustin) [inline client/src/island/island-state-command.ts:202]:
```suggestion
                if (!islandPattern) {
                    return
                }
                ...
```

File: client/src/storage/storage.ts [hunk 10]
```diff
@@ -1,24 +1,20 @@
 import {IGenomeData} from "../genetics/genome"
 import {Hexalot} from "../island/hexalot"
-import { Island, IslandPattern } from "../island/island"
-import {Journey} from "../island/journey"
+import {Island, IslandData} from "../island/island"
+import {IJourneyData} from "../island/journey"
 
 export interface IStorage {
```
Comment (mitchelljustin) [inline client/src/storage/storage.ts:6]: Might want to rename this something like `IAppStorage` to disambiguate

**Review by mitchelljustin — APPROVED** [review]

File: client/src/island/hexalot.ts (hunk 5 above)
Comment (geralddejong) [reply client/src/island/hexalot.ts]: did it.

File: client/src/island/hexalot.ts (hunk 8 above)
Comment (geralddejong) [reply client/src/island/hexalot.ts:187]: yes

File: client/src/storage/storage.ts (hunk 10 above)
Comment (geralddejong) [reply client/src/storage/storage.ts:6]: I don't see ambiguity yet.

File: client/src/view/island-component.tsx [hunk 11]
```diff
@@ -120,27 +120,36 @@ export class IslandComponent extends React.Component<IslandComponentProps, objec
...
         })
+        return geometry
+    }
+
+    private get vacantHangersGeometry(): Geometry {
+        const vacantHexalot = this.props.islandState.island.vacantHexalot
         const geometry = new Geometry()
-        hexalots.forEach(hexalot => hexalot.centerSpot.addHangerGeometry(geometry.vertices))
-        geometry.computeBoundingSphere()
+        // if (vacantHexalot) { // why does a disposed one get included?
```
Comment (mitchelljustin) [inline client/src/view/island-component.tsx:138]: Remove?

**Comment by geralddejong** [comment]
did it in the other repo

**Pull request closed without merging**